File: admin/src/lib/usage.ts
// usage.ts — REGISTRO DE GASTOS de APIs (Centro de Gastos), lado TypeScript.
// recordUsage(...) añade una línea a data/usage.jsonl calculando el USD según data/pricing.json.
// Lo usan los generadores TS (Veo en faceless/sabikids, gpt-image-1). Nunca lanza: si falla, no rompe nada.
import fs from 'node:fs';
import path from 'node:path';

const ADMIN = process.cwd(); // los scripts del admin corren con cwd = raíz del admin
const USAGEF = path.join(ADMIN, 'data', 'usage.jsonl');
const PRICEF = path.join(ADMIN, 'data', 'pricing.json');

function pricing(): any {
  try { return JSON.parse(fs.readFileSync(PRICEF, 'utf8')); } catch { return {}; }
}
function rate(provider: string, model: string): any {
  const pr = pricing()[provider] || {};
  return pr[model] || pr.default || {};
}

export function recordUsage(
  project: string, kind: string, provider: string, model: string,
  units: number, usd?: number, note?: string,
): number {
  const r = rate(provider, model);
  const u = usd == null ? Number(((units || 0) * (r.usd || 0)).toFixed(6)) : usd;
  const line = {
    ts: new Date().toISOString(), project: project || '?', kind, provider, model,
    units: Number((units || 0).toFixed(3)), unit: r.per || 'unit',
    usd: Number((u || 0).toFixed(6)), note: String(note || '').slice(0, 120),
  };
  try { fs.mkdirSync(path.dirname(USAGEF), { recursive: true }); fs.appendFileSync(USAGEF, JSON.stringify(line) + '\n'); } catch { /* noop */ }
  return line.usd;
}

export function recordTokens(project: string, kind: string, model: string, tin: number, tout: number, note?: string): number {
  const r = rate('openai', model);
  const usd = Number(((tin / 1000) * (r.usd_in || 0) + (tout / 1000) * (r.usd || 0)).toFixed(6));
  return recordUsage(project, kind, 'openai', model, (tin || 0) + (tout || 0), usd, note);
}

File: admin/py/usage.py
"""usage.py — REGISTRO DE GASTOS de APIs (Centro de Gastos).
record(...) añade una línea a data/usage.jsonl y calcula el USD según data/pricing.json.
Lo usan todos los generadores (Veo, OpenAI gpt-image / gpt-4o, ElevenLabs, Whisper).
Nunca lanza excepción: si algo falla, no rompe la generación.
"""
import os, json, datetime
ADMIN = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
USAGEF = os.path.join(ADMIN, "data", "usage.jsonl")
PRICEF = os.path.join(ADMIN, "data", "pricing.json")

def _pricing():
    try: return json.load(open(PRICEF, encoding="utf-8"))
    except Exception: return {}

def _rate(provider, model):
    pr = _pricing().get(provider, {}) or {}
    return pr.get(model) or pr.get("default") or {}

def record(project, kind, provider, model, units, usd=None, note=""):
    r = _rate(provider, model)
    try:
        if usd is None: usd = float(units) * float(r.get("usd", 0))
    except Exception: usd = 0
    line = {"ts": datetime.datetime.now(datetime.timezone.utc).isoformat(),
            "project": project or "?", "kind": kind, "provider": provider, "model": model,
            "units": round(float(units or 0), 3), "unit": r.get("per", "unit"),
            "usd": round(float(usd or 0), 6), "note": str(note)[:120]}
    try:
        os.makedirs(os.path.dirname(USAGEF), exist_ok=True)
        with open(USAGEF, "a", encoding="utf-8") as f: f.write(json.dumps(line, ensure_ascii=False) + "\n")
    except Exception: pass
    return line["usd"]

def record_tokens(project, kind, model, tin, tout, note=""):
    r = _rate("openai", model)
    try: usd = (tin / 1000.0) * float(r.get("usd_in", 0)) + (tout / 1000.0) * float(r.get("usd", 0))
    except Exception: usd = 0
    return record(project, kind, "openai", model, (tin or 0) + (tout or 0), usd=usd, note=note)
